### PR TITLE
Determine the fork from the Git environment.

### PR DESCRIPTION
**Pull request by netdpb**

### CODE REVIEW

**Comment by netdpb** [comment]
Why didn't the build action run?

**Comment by cpovirk** [comment]
I saw an email half an hour ago about some kind of partial GitHub outage, including impact on GitHub Actions. So probably that.

https://www.githubstatus.com/incidents/smn1qtqvbsb6

**Comment by netdpb** [comment]
Let's not merge this until the [issue](https://www.githubstatus.com/incidents/smn1qtqvbsb6) with GitHub Actions is resolved.

**Comment by netdpb** [comment]
Let's not merge until the issue is resolved.

File: initialize-project [hunk 1]
```diff
@@ -29,6 +30,16 @@ run() {
   "$@"
 }
 
+forked_repo() {
+  if [[ -n "${FORK:-}" ]]; then
+    echo "${FORK}"
+    return
+  fi
+  local url
+  url="$(git config remote.origin.url --default '')"
```
Comment (cpovirk) [inline initialize-project:39]: If I clone `git@github.com:cpovirk/nullness-checker-for-checker-framework.git` (or `https://github.com/cpovirk/nullness-checker-for-checker-framework.git`), this `git config` command produces empty output when run in the resulting directory. Is that expected? Maybe what you have here works for CI set up in your fork? I know nothing about `remote.origin.url`, so I'd naively assumed it was "the thing you passed to `git clone`.")

Comment (netdpb) [reply initialize-project:39]: The default name for the remote used to clone is "origin".

What command did you use to clone?

Comment (cpovirk) [reply initialize-project:39]: `git clone git@github.com:cpovirk/nullness-checker-for-checker-framework.git`

`git clone https://github.com/cpovirk/nullness-checker-for-checker-framework.git`

I just tried cloning and running the command again after removing my `$HOME/.gitconfig`, and I still see the empty command output.

Comment (cpovirk) [reply initialize-project:39]: The repo's `.git/config` doesn't seem to have a record of the URL anywhere??

```
$ cat .git/config
[core]
	repositoryformatversion = 0
	filemode = true
	bare = false
	logallrefupdates = true
[remote "origin"]
	url = --default
	fetch = +refs/heads/*:refs/remotes/origin/*
[branch "main"]
	remote = origin
	merge = refs/heads/main
```

```
$ git push -n
fatal: strange pathname '--default' blocked
```

Uh... did something break Git on my machine?

Comment (cpovirk) [reply initialize-project:39]: Oh, sorry, that's what I see _only after I run `git config remote.origin.url --default ''`_.

Before then, I see what I'd expect:
```
	url = git@github.com:cpovirk/nullness-checker-for-checker-framework.git
```

So the `git config` command appears to be modifying the value. Is this one of those annoying argument-ordering things, where `--default` needs to come before the positional args?

Comment (cpovirk) [reply initialize-project:39]: > Is this one of those annoying argument-ordering things, where `--default` needs to come before the positional args?

Yeah, probably:

```
$ git config --default '' remote.origin.url
git@github.com:cpovirk/nullness-checker-for-checker-framework.git
```

Comment (netdpb) [reply initialize-project:39]: Hahaha I know what it is.

The default is `--list`, so I left it out, which worked until I added another option.

File: initialize-project [hunk 2]
```diff
@@ -10,17 +10,18 @@
 #
 # Set SHALLOW=1 to clone sibling projects at depth 1.
 #
-# Set FORK to the GitHub organization that holds your fork of
-# jspecify/checker-framework, jspecify/jspecify, or jspecify/jdk.
+# This script automatically tries to download your fork of
+# jspecify/checker-framework, jspecify/jspecify, or jspecify/jdk, if they exist.
+# It looks at remote.origin.url to determine that.
```
Comment (cpovirk) [inline initialize-project:15]: For people like me, how about something like:

"(Normally, remote.origin.url is set to the URL you passed to \`git clone\`.)"

File: initialize-project (hunk 1 above)
Comment (cpovirk) [reply initialize-project]: Still not quite?

```
$ git config --get remote.origin.url --default ''
error: wrong number of arguments, should be from 1 to 2
```

I see that locally and in https://github.com/jspecify/nullness-checker-for-checker-framework/actions/runs/3191280029/jobs/5207401018.

(I have some horrible boilerplate that I use to try to get Bash to exit even for errors inside `$(...)` and so forth (`set -E; trap 'exit 1' USR1; trap 'kill -USR1 $$; pkill -P $$' ERR`, maybe also the function that I override `exit` with?), but I very much don't _recommend_ it... :))

**Comment by netdpb** [comment]
Oh for @#$%^&* sake. The `--default` has to come before the `--get`.

File: initialize-project [hunk 3]
```diff
@@ -29,6 +31,16 @@ run() {
...
 }
 
+forked_repo() {
+  if [[ -n "${FORK:-}" ]]; then
+    echo "${FORK}"
+    return
+  fi
+  local url
+  url="$(git config --default '' --get remote.origin.url)"
+  echo "${url%/$(git branch --show-current)}"
```
Comment (cpovirk) [inline initialize-project:41]: What sort of clone includes the branch name in the output of the `git config` command? I don't see it for a clone like `git clone git@github.com:jspecify/jspecify.git -b samples-google-prototype`, but I didn't try any fancy settings.

I would also accept "The current script works, and I don't want to think about this anymore" :) But if you already know, and if it's trivial to educate me on this piece of Git, I'm interested.

File: initialize-project [hunk 4]
```diff
@@ -43,8 +55,10 @@ git_clone() {
   fi
   git+=("$@")
 
-  if [[ -n "${FORK:-}" ]]; then
-    if run "${git[@]}" "git@github.com:${FORK}/${repo}.git" "../${repo}"; then
+  local fork
+  fork="$(forked_repo)"
+  if [[ -n "${fork}" ]]; then
+    if run "${git[@]}" "${fork}/${repo}.git" "../${repo}"; then
```
Comment (cpovirk) [inline initialize-project:61]: It looks like we need to strip the name of the `nullness-checker-for-checker-framework` repo from `${fork}`. Possibly that's even what you were looking to do with the `branch` stuff above???

File: initialize-project (hunk 3 above)
Comment (netdpb) [reply initialize-project:41]: Hmm. You're right. I am trying to chop off the name of the current repo, but that's not the branch.

File: initialize-project (hunk 4 above)
Comment (netdpb) [reply initialize-project]: Yes. What I have now should do it.

**Comment by cpovirk** [comment]
I've got to hand it to you, using `dirname` is pretty cute :)

**Pull request merged**